lib/f_service/result/base.rb

```diff
@@ -75,7 +82,13 @@ def on(success:, failure:)
       # @return [Success, Failure] the original Result object
       # @api public
       def on_success(*target_types)
-        yield(*to_ary) if successful? && expected_type?(target_types)
+        old_callback_any_matching_warn(method_name: __method__, from: caller[0]) if target_types.empty?
```

Comment:
Can you clarify why are we warning on this? I think having a plain `on_success { do_x }/on_failure { do_x }` is okay, since not always we will use result types.

Reply:
Well, the problem we were wanting to solve here was exactly that...

- The callback propagation never stops even if some callback processes the result;
- The callback without args matches everything;

Then if we run this...

```rb
User::Update.(user: user)
  .on_success { |value| return json_success(value) }
  .on_failure(:error_1) { |value| return catch_error_1(value) }
  .on_failure(:error_2) { |value| return catch_error_2(value) }
  .on_failure { |value| return catch_any_unknown_error(value) }
```

Even if error 1 or 2 is processed, the unknown error will be processed as well.

Then, the PR proposed just to add a stop propagation to if error 1 or 2 process the result the * matcher does not process anything...

Then, you suggested (if I understood well), to change the `on_(failure/success)` to add an argument to indicate that it matches anything unhandled until that callback runs.

Then, seems that does not make sense to have 2 ways to process any unhandled result, because...

```rb
User::Update.(user: user)
  .on_success { |value| return json_success(value) }
  .on_failure(:error_1) { |value| return catch_error_1(value) }
  .on_failure(:error_2) { |value| return catch_error_2(value) }
  .on_failure { |value| return catch_any_unknown_error(value) } # Process any unhandled result
  .on_failure(unhandled: true) { |value| return catch_any_unknown_error(value) } # Will never run

User::Update.(user: user)
  .on_success { |value| return json_success(value) }
  .on_failure(:error_1) { |value| return catch_error_1(value) }
  .on_failure(:error_2) { |value| return catch_error_2(value) }
  .on_failure(unhandled: true) { |value| return catch_any_unknown_error(value) } # Process any unhandled result
  .on_failure { |value| return catch_any_unknown_error(value) } # Will never run also
```

This, to me, could result in more confusion than let as it was and use just as it was (passing no arguments)...

That's why without any args becomes deprecated.

Reply:
I still think it's valuable to have a plain `on_success/on_failure` without a type. Many interactions only have two flows: failure and success, so requiring a type is cumbersome and error-prone (imagine misspelling an error type `.on_failure(:user_not_fonud)`)

One alternative would be making this configurable Gem-level: i.e., allowing users to define if `on_success/on_failure` should run only on unhandled failures or not.

```ruby
FService.config do |config|
  config.callbacks_run_only_on_unhandled_results = true
end
```

I think this is a bit too much to my taste, but I imagine y'all ran into some kind of bug due to this.